File: libs/shared-py/auctiondeal_shared/telemetry.py
"""Telemetry and logging utilities for Python services."""

import os
import time
from typing import Any, Dict, Optional
import structlog
from structlog import get_logger


# Configure structured logging
structlog.configure(
    processors=[
        structlog.stdlib.filter_by_level,
        structlog.stdlib.add_logger_name,
        structlog.stdlib.add_log_level,
        structlog.stdlib.PositionalArgumentsFormatter(),
        structlog.processors.TimeStamper(fmt="iso"),
        structlog.processors.StackInfoRenderer(),
        structlog.processors.format_exc_info,
        structlog.processors.UnicodeDecoder(),
        structlog.processors.JSONRenderer() if os.getenv('LOG_FORMAT') == 'json' else structlog.dev.ConsoleRenderer(),
    ],
    context_class=dict,
    logger_factory=structlog.stdlib.LoggerFactory(),
    cache_logger_on_first_use=True,
)

logger = get_logger()


class TelemetryUtils:
    """Utility functions for telemetry and monitoring."""
    
    @staticmethod
    def start_timer(operation: str, context: Optional[Dict[str, Any]] = None) -> callable:
        """Start timing an operation."""
        start_time = time.time()
        logger.info("Starting operation", operation=operation, **(context or {}))
        
        def end_timer() -> float:
            duration = time.time() - start_time
            logger.info("Completed operation", operation=operation, duration=duration, **(context or {}))
            return duration
        
        return end_timer
    
    @staticmethod
    def log_error(error: Exception, context: Optional[Dict[str, Any]] = None) -> None:
        """Log error with structured context."""
        logger.error(
            "Error occurred",
            error=str(error),
            error_type=type(error).__name__,
            **(context or {})
        )
    
    @staticmethod
    def log_info(message: str, **kwargs) -> None:
        """Log info message with context."""
        logger.info(message, **kwargs)
    
    @staticmethod
    def log_debug(message: str, **kwargs) -> None:
        """Log debug message with context."""
        logger.debug(message, **kwargs)
    
    @staticmethod
    def log_warning(message: str, **kwargs) -> None:
        """Log warning message with context."""
        logger.warning(message, **kwargs)

File: libs/shared-ts/src/telemetry.ts
import { createLogger, format, transports } from 'winston'

// Create structured logger
export const logger = createLogger({
  level: process.env.LOG_LEVEL || 'info',
  format: format.combine(
    format.timestamp(),
    format.errors({ stack: true }),
    format.json()
  ),
  transports: [
    new transports.Console({
      format: process.env.NODE_ENV === 'development'
        ? format.combine(format.colorize(), format.simple())
        : format.json()
    })
  ]
})

// Performance monitoring
export class TelemetryUtils {
  static startTimer(operation: string): () => number {
    const start = Date.now()
    logger.info(`Starting operation: ${operation}`)
    
    return () => {
      const duration = Date.now() - start
      logger.info(`Completed operation: ${operation}`, { duration })
      return duration
    }
  }

  static logError(error: Error, context?: Record<string, unknown>): void {
    logger.error('Error occurred', {
      error: error.message,
      stack: error.stack,
      ...context
    })
  }

  static logInfo(message: string, context?: Record<string, unknown>): void {
    logger.info(message, context)
  }

  static logDebug(message: string, context?: Record<string, unknown>): void {
    logger.debug(message, context)
  }
}